persistence-api/src/main/java/io/stargate/db/Persistence.java

```diff
@@ -15,96 +15,162 @@
  */
 package io.stargate.db;
 
+import com.datastax.oss.driver.shaded.guava.common.util.concurrent.Uninterruptibles;
 import io.stargate.db.datastore.DataStore;
 import io.stargate.db.schema.Schema;
-import java.net.InetSocketAddress;
-import java.net.SocketAddress;
 import java.nio.ByteBuffer;
 import java.util.List;
-import java.util.Map;
+import java.util.Optional;
 import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+import org.apache.cassandra.stargate.exceptions.AuthenticationException;
 import org.apache.cassandra.stargate.locator.InetAddressAndPort;
-import org.apache.cassandra.stargate.utils.MD5Digest;
 
-public interface Persistence<T, C, Q> {
-  String name();
+/**
+ * A persistence layer that can be queried.
+ *
+ * <p>This is the interface that stargate API extensions uses (either directly, or through the
+ * higher level {@link DataStore} API which wraps a instance of this interface) to query the
+ * underlying store, and thus the one interface that persistence extensions must implement.
+ */
+public interface Persistence {
 
-  void initialize(T config);
+  /** Name describing the persistence implementation. */
+  String name();
 
-  void destroy();
+  /** The current schema. */
+  Schema schema();
 
   /**
-   * Returns the current schema.
-   *
-   * @return The current schema.
+   * Registers a listener whose methods will be called when either the schema or the topology of the
+   * persistence changes.
    */
-  Schema schema();
-
   void registerEventListener(EventListener listener);
 
   boolean isRpcReady(InetAddressAndPort endpoint);
 
   InetAddressAndPort getNativeAddress(InetAddressAndPort endpoint);
 
-  QueryState<Q> newQueryState(ClientState<C> clientState);
-
-  ClientState<C> newClientState(SocketAddress remoteAddress, InetSocketAddress publicAddress);
-
-  ClientState newClientState(String name);
-
-  AuthenticatedUser<?> newAuthenticatedUser(String name);
-
   Authenticator getAuthenticator();
 
-  DataStore newDataStore(QueryState<Q> state, QueryOptions queryOptions);
+  /**
+   * Creates a new connection for an "external" client identified by the provided info.
+   *
+   * @param clientInfo info identifying the client for which the connection is created.
+   * @return the newly created persistence connection.
+   */
+  Connection newConnection(ClientInfo clientInfo);
```

Comment:
`ClientInfo` appears to require an `InetSocketAddress`... it will be difficult to provide for SQL clients, for example (because socket info is hidden behind Avatica driver APIs).

How critical is having the client's address available?

Reply:
> `ClientInfo` appears to require an `InetSocketAddress`

First, I'll note it's not really new to this patch (it was before in the stargate `ClientState` wrapper, that is replace by this `ClientInfo`). It's due to C* underlying `ClientState` class using `InetSocketAddress`.

Now, why does C* cares about he client address you may ask? It's a good question, and I'd need to check in details. But at a minimum, I believe it uses it in some logging/audit logging, and possibly tracing.

But I wouldn't worry for SQL. Simplest solution is we just don't provide a `ClientInfo` at all (check `newConnection`, there is an argument-less version), and that's likely just fine. And maybe longer term we figure out it's neat to have "some" client info for SQL, and I'm sure we can figure out a way to do it (we can hardcode the port for instance).